mlperf_logging/package_checker/seed_checker.py

```diff
@@ -0,0 +1,148 @@
+        path: The absolute or relative path to the directory under query.
+    """
+    source_files = []
+    for root, subdirs, files in os.walk(path):
+        for file_name in files:
+            if is_source_file(file_name):
+                source_files.append(os.path.join(root, file_name))
+    return source_files
+
+
+class SeedChecker:
+    """ Check if the seeds fit MLPerf submission requirements.
+    Current requirements are:
+
+    1. All seeds must be logged through mllog (if choose to log seeds). Any seed
+       logged via any other method will be disgarded.
+    2. All seeds, if choose to be logged, must be valid integer (convertible via
+       int()).
+    3. If any run log at lesat one seed, we expect all runs to log at least
+       one seed.
+    4. The set of seed(s) that one run logs must be completely different from
+       the set of seed(s) any other run logs.
+
+    Unsatisfying any of the above requirements results in check failure.
+
+    A warning is raised for the following situations:
+
+    1. Any run logs more than one seed.
+    2. No seed is logged. However, the source files contain the keyword whose
+       lowercase is "seed". What files are considered as source files are
+       defined in SOURCE_FILE_EXT and is_source_file().
+
+    """
+    def __init__(self, ruleset):
+        self._ruleset = ruleset
+
+    def _get_seed(self, result_file):
+        loglines, errors = mlp_parser.parse_file(
+            result_file,
+            ruleset=self._ruleset,
+        )
+        if len(errors) > 0:
+            raise ValueError('\n'.join(
+                ['Found parsing errors:'] +
+                ['{}\n  ^^  {}'.format(line, error)
+                 for line, error in errors] +
+                ['', 'Log lines had parsing errors.']))
+        return [
+            int(line.value['value']) for line in loglines if line.key == 'seed'
+        ]
+
+    def _assert_unique_seed_per_run(self, result_files):
+        no_logged_seed = True
+        error_messages = []
+        seed_to_result_file = {}
+        for result_file in result_files:
+            try:
+                seeds = self._get_seed(result_file)
+            except Exception as e:
+                error_messages.append("Error found when querying seeds from "
+                                      "{}: {}".format(result_file, e))
+                continue
+
+            if not no_logged_seed and len(seeds) == 0:
+                error_messages.append(
+                    "Result file {} logs no seed. However, other "
+                    "result files, including {}, already logs some "
+                    "seeds.".format(list(seed_to_result_file.keys())))
+            if no_logged_seed and len(seeds) > 0:
+                no_logged_seed = False
+            if len(seeds) > 1:
+                warnings.warn("Result file {} logs more than one "
+                              "seeds {}!".format(result_file, seeds))
+            for seed in seeds:
```

Comment:
Here, if result files F1 and F2 have seeds S1 and S2, and F1.S1 == F2.S2 this will be marked as a violation even though it is not, right? This case is rare, but do we handle it?

Reply:
By S1 and S2, you meant seeds logged at different lines between different result files?

Reply:
Yes, so each file has S1 and S2, and S1 and S2 are the same for different files. Not sure if it is a valid scenario

Reply:
I updated a few new changes to support this. I think what you pointed out is a good idea in general, however, currently I don't think we have a way to make sure that what being logged matches the source code, thus, seed assertion on a file/lineno granularity might be not rigorous enough. For example, say I launch one run, then I modified something minor in the code (in the sense that I subjectively think a re-run is not really needed) and then launch a second run; even if I use the same seed, the file/lineno is going to be different.